### ISSUE TITLE
Support for `sklearn.preprocessing.FunctionTransformer`

### ISSUE THREAD

**Issue by jrwalk** [jpmml/jpmml-sklearn#10]
Would it be possible to implement the `FunctionTransformer` functionality in the supported preprocessing tools?  it currently plays well with the pickle serialization of the `DataFrameMapper` object, but is failing on the call to the jar.  I imagine the full functionality would be difficult due to the limitations of supported functions in the PMML format, but at the least some common transforms wrapped in the FunctionTransformer (e.g., basic numerical operations) are supported in PMML.

**Comment by vruusmann** [comment]
I have it in my TODO list.

It's good to hear that the `FunctionTransformer` class is serializable. I wonder how the expression is stored - is it a `string`, or some pre-parsed data structure?

It would be quite a lot of work to build a parser for the `string` representation of expressions. Perhaps it would be possible to leverage some other Java-to-Python framework such as Jython for this.

**Comment by jrwalk** [comment]
That's encouraging to hear!

The serialization seemed to work fine, although mine was a fairly simplistic case -- the `FunctionTransformer` is simply wrapping around some `numpy` functions (thus my thinking they can be implemented here, as I know at least for my case the mathematical ops involved are supported in the PMML format).  However, this may be difficult in the general case -- looking at the serialization, it seems the pickle is just storing a string reference to the base `numpy` implementation.

**Comment by jrwalk** [comment]
Course the workflow is still getting hiccups due to the fact that scikit-learn's `svm.OneClassSVM` isn't supported, but that's an entirely separate issue (and one that's fundamental to PMML I think, not your API)!

**Comment by vruusmann** [comment]
The `OneClassSVM` model type should be moved to a separate issue.

At first glance, I don't see anything too difficult about it. It's the same LibSVM data structure which is already in use by `(Nu)SV(C|R)` model types. The challenge is to compose a thin wrapper around the `SupportVectorMachineModel` PMML element - has to be some classification-type function that binarizes the raw SVM score to "+1" or "-1" class label.

**Comment by jrwalk** [comment]
ha, yeah I'd figured that would be a deeper issue with the PMML support so I hadn't made it a formal issue.  I'll kick that out now.

**Comment by vruusmann** [comment]
This is a very limited implementation, which currently supports [Numpy universal functions] (http://docs.scipy.org/doc/numpy/reference/ufuncs.html) `absolute`, `exp`, `log`, `log10` and `sqrt`. It can be extended to support other single-argument functions such as `exp2`, `log2`, `log1p` etc. if needed.

However, any idea how to invoke multi-argument functions such as `add` or `subtract` in the context of a `DataFrameMapper`? Suppose that I'm interested in adding a constant `1` to all Iris dataset argument fields?

**Comment by jrwalk** [comment]
Awesome!  Just covering the numpy universals goes a long way to covering typical use cases, I think.

As for multi-argument functions, I'm not sure I'm the right person to ask, as I'm fairly new to this workflow.  However, my gut feeling would be that this isn't easily implementable -- for the `FunctionTransformer`, the clearest way I can see would be setting the transformer with the `pass_y` flag as `True`, then passing in the second argument as the (nominal) y-value to in the call to `FunctionTransformer.transform`, as this just forwards the parameters to the stored `func` (a moderately evil hack).  However, calling a transformer in `DataFrameMapper` with multiple columns expects that to be a two-dimensional array for X, rather than separating them into X- and y-values.  Adding a constant may be simpler than this case though.
